atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1114,24 +1114,27 @@ private int estimateSize(List<RowResult<Value>> rangeRows) {
                 } else {
                     // The value has a commit timestamp less than our start timestamp, and is visible and valid.
                     if (value.getContents().length != 0) {
-                        found++;
                         results.put(key, transformer.apply(value));
+                        keysAddedBuilder.add(key);
                     }
                 }
             }
         }
-        count.set(found);
+        Set<Cell> keysAddedToResults = keysAddedBuilder.build();
+        count.addAndGet(keysAddedToResults.size());
 
         if (!keysToDelete.isEmpty()) {
             // if we can't roll back the failed transactions, we should just try again
             if (!rollbackFailedTransactions(tableRef, keysToDelete, commitTimestamps, defaultTransactionService)) {
+                rawResults.keySet().removeAll(keysAddedToResults);
```

Comment:
So... this is a little tricky, because it seems to me at least that `rawResults` may be an ImmutableMap and so a removeAll operation wouldn't be supported.

https://github.com/palantir/atlasdb/blob/8f526dd10790601199292f948dff3507d5830b68/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java#L433

passed into getRowColumnRangePostFiltered
passed into this.

---

One way is to return a view of the map by doing filterKeys on the key not being in this set. If we need to retry a lot that may not be such a good idea. Alternatives could be
1. take a mutable copy at all call-sites (which is fine, because this method is private)
2. make a copy on each iteration

Reply:
Yeah, you are right, if that is hitting my changes then the call will fail. Why then, are we using the `Map` interface in all the places? Should we replace that with `ImmutableMap` where it may be the case?